Linear search for Linked List

BengRhui opened the pull request:
The method to add linear search for linked list is added. @cmchan04 you may test it with this code. Make sure to place it after the linked list is created ya.

```
    // Attempt to search for a specific transition type
    int searchSize = 0;
    string transactionName = "deposit";
    cout << endl << "Try searching for deposit: " << endl;
    Transaction** searchedTransaction = Searcher::linearSearchUsingList(*list, transactionName, searchSize);

    // Print details of search
    cout << "Size: " << searchSize << endl;
    if (searchedTransaction == nullptr) cout << "No transactions found." << endl;
    else {
        cout << "Example of three transactions are: " << endl;
        for (int i = 0; i < 3; i++) {
            cout << "Transaction Number " << i << endl;
            cout << "  ID: " << searchedTransaction[i] -> transactionId << endl;
            cout << "  Payment Channel: " << searchedTransaction[i] -> paymentChannel << endl;
            cout << "  Transaction Type: " << searchedTransaction[i] -> transactionType << endl;
        }
    }
```

BengRhui committed:
Add method to calculate current memory used

The method to calculate the current memory used in macOS is added. This is not tested in Windows yet, so additional changes might be made.

BengRhui committed:
Update performance tracker on windows

The print method is updated to display beginning, final and their difference.

BengRhui committed:
Move Node struct outside of class

The Node struct is moved outside of the DoublyLinkedList class to make sure that it can be accessed in other classes.

BengRhui committed:
Add linear search for linked list

The static method for linear search is added.

cmchan04 reviewed (commented):
LGTM. I have included a toLowerCase function in Searcher in the latest commit. Maybe you can use it also❤️❤️

cmchan04 reviewed (changes requested):
![image](https://github.com/user-attachments/assets/4682b79f-6748-422d-a617-e492a7076c55)
The ranges::transform is showing an error, maybe forget to include ```<algorithm>``` and ```<ranges>``` in ```Searcher.h```

cmchan04 commented:
Others LGTM. 😊😊

BengRhui commented:
> ![image](https://private-user-images.githubusercontent.com/162660825/457400927-4682b79f-6748-422d-a617-e492a7076c55.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTA0OTU0NjQsIm5iZiI6MTc1MDQ5NTE2NCwicGF0aCI6Ii8xNjI2NjA4MjUvNDU3NDAwOTI3LTQ2ODJiNzlmLTY3NDgtNDIyZC1hNjE3LWU0OTJhNzA3NmM1NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNjIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDYyMVQwODM5MjRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lM2MwZTJhMGUwODQwYmI2OTM2MjNlNjhmMWFlNWM0ODJkZWZmMzBmNDdkMDkwNTljNmMwOTM3ZmU4MDkxNGQxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.fL7W9SDA5EGu2D6FZAtpHsUTF89No1m-6yK3jQ89GA8) The ranges::transform is showing an error, maybe forget to include `<algorithm>` and `<ranges>` in `Searcher.h`

Weird, this doesn't show any error on my side. I shall try running it on Parallels and see.

BengRhui reviewed (commented):
@cmchan04 On my side, it seems that everything is running without any issues, both on Mac and Windows. Maybe you shall retry on your side and see if there's any compilation-related issues (maybe try rebuild and run the project and see)?

On Parallels:
<img width="1481" alt="Screenshot 2025-06-21 at 4 45 37 PM" src="https://github.com/user-attachments/assets/c5ac7837-925f-4f55-8af3-33e3fa6d03d3" />

On Mac:
<img width="1351" alt="Screenshot 2025-06-21 at 4 50 03 PM" src="https://github.com/user-attachments/assets/ac0f420f-62d3-4fd2-ac9a-df752b4e7600" />

BengRhui committed:
Copy constructor and insert before

Additional methods are added into the linked list class

BengRhui committed:
Code refactoring on shrinking list size

The method that shrinks a list with empty spaces is separated from the main linear search algorithm.

BengRhui commented:
Btw @cmchan04 I've added a few additional commits to separate some methods out from my linear search algorithm. Do let me know if there's anything wrong with the commits ya

cmchan04 commented:
> @cmchan04 On my side, it seems that everything is running without any issues, both on Mac and Windows. Maybe you shall retry on your side and see if there's any compilation-related issues (maybe try rebuild and run the project and see)?
> 
> On Parallels: <img alt="Screenshot 2025-06-21 at 4 45 37 PM" width="1481" src="https://private-user-images.githubusercontent.com/121010231/457565648-c5ac7837-925f-4f55-8af3-33e3fa6d03d3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTA1MDA2OTMsIm5iZiI6MTc1MDUwMDM5MywicGF0aCI6Ii8xMjEwMTAyMzEvNDU3NTY1NjQ4LWM1YWM3ODM3LTkyNWYtNGY1NS04YWYzLTMzZTNmYTZkMDNkMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNjIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDYyMVQxMDA2MzNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yODQzNWMxMjEyYWE2MGM1M2MxYjJlYzYwZTQ3NGMxZTk2ZjVjODdiNjA1Y2YzZDA0NDcwYjM5ZGVjNDNjNTlmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.3dP4EcbZNVn2byTip0FuW49732q9JBVvpBs-cqNOUMg">
> 
> On Mac: <img alt="Screenshot 2025-06-21 at 4 50 03 PM" width="1351" src="https://private-user-images.githubusercontent.com/121010231/457565896-ac0f420f-62d3-4fd2-ac9a-df752b4e7600.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTA1MDA2OTMsIm5iZiI6MTc1MDUwMDM5MywicGF0aCI6Ii8xMjEwMTAyMzEvNDU3NTY1ODk2LWFjMGY0MjBmLTYyZDMtNGZkMi1hYzlhLWRmNzUyYjRlNzYwMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNjIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDYyMVQxMDA2MzNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03MmFmNDk4NTgzOGYxOTQzMzFhZTQ2NjYzMzY3M2EwNWRjOGZmNGZlMDc3ZTE5MjU2NmVlMmUyZTk2MWQyNmUxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.pGJ2LMZFZPBF-omQFfN2bdZ6WAZ5ipagti5ZZd9gD4E">

![image](https://github.com/user-attachments/assets/e829712b-8fbb-4eef-a5ef-5ba8caa690f0)
Here's how it runs without include algorithm

Then here is the result with 
![image](https://github.com/user-attachments/assets/4f4baa33-8c48-4c24-afd5-da72905e6eef)
I am pretty sure it is becuz of the library not included lah, cuz I am also using the clang c++ 20 version, will including these two library affect the performance on ur side @BengRhui ?

PS: I tried without ranges, it works, but can't without algorithm.

BengRhui commented:
Including library won't affect performances one, don't worry @cmchan04. If that's the case, let's just include both `#include <ranges>` and `#include <algorithm>`inside our code to prevent any compilation errors.

If everything else looks good, maybe you can add a commit to include the imports and merge this branch to main?

cmchan04 reviewed (changes requested):
There are some code block remained after the ```removeUnusedIndex``` function on line 103 - 110

cmchan04 committed:
include algorithm in Searcher, remove extra code blocks
